### PR TITLE
[WIRES] Add tests for nonconsecutive wires

### PR DESCRIPTION
**Pull request by mariaschuld**
**Context:**

This is the last of the wire management refactor PRs. It merges into the ``allow_non_consecutive_wires`` branch, which carries all (breaking) changes ensuing from allowing devices to have registers with arbitrary wire labels.

This PR adds tests to different places of the PennyLane code base.

**Description of the Change:**

The basic logic of the tests is this:

* wire management tests are written as integration tests
* where possible, they compare the results of qnodes run with different wire labels with each other
* the tests are added to the ``device`` and ``qubit_device`` test files, as well to templates, circuit graph, circuit drawer and core device tests.

### CODE REVIEW

File: pennylane/plugins/default_qubit.py [hunk 1]
```diff
@@ -334,12 +334,5 @@ def analytic_probability(self, wires=None):
         if self._state is None:
             return None
```
Comment (mariaschuld) [inline pennylane/plugins/default_qubit.py:336]: This was a bug oncovered by the tests: this function should rather just pass on the wires...

File: tests/circuit_graph/test_circuit_graph.py [hunk 2]
```diff
@@ -61,24 +61,31 @@ def circuit(ops):
 
 
 @pytest.fixture
-def parameterized_circuit():
+def wires():
+    """A fixture for a mixed-type wires label list of length 3."""
+    return ['a', 'q1', 3]
+
+
```
Comment (mariaschuld) [inline tests/circuit_graph/test_circuit_graph.py:68]: This runs the test on the "interesting" wires...

File: tests/test_qubit_device.py [hunk 3]
```diff
@@ -583,6 +581,10 @@ class TestMarginalProb:
             ([0, 2], [1]),
             ([1, 2], [0]),
             ([0, 1, 2], []),
```
Comment (mariaschuld) [inline tests/test_qubit_device.py:583]: Just adding some examples with ``Wires`` input to check if the function takes them.

File: tests/beta/test_default_tensor.py [hunk 4]
```diff
@@ -1754,3 +1754,31 @@ def test_hermitian(self, theta, phi, varphi, monkeypatch, rep, tol):
...
+
+    @pytest.mark.parametrize("wires1, wires2", [(['a', 'c', 'd'], [2, 3, 0]),
+                                                ([-1, -2, -3], ['q1', 'ancilla', 2])])
+    def test_wires_expval(self, wires1, wires2, tol):
+        """Test that the expectation of a circuit is independent from the wire labels used."""
+
+        circuit1 = self.make_circuit_expval(wires1)
+        circuit2 = self.make_circuit_expval(wires2)
+
+        assert np.allclose(circuit1(), circuit2(), tol)
```
Comment (josh146) [inline tests/beta/test_default_tensor.py:1784]: Nice! does this also work passing the wires as QNode arguments?

```python
@qml.qnode(dev)
def circuit(wires):
    qml.RX(0.5, wires=wires[0])
    qml.RY(2., wires=wires[1])
    qml.CNOT(wires=[wires[0], wires[2]])
    qml.CNOT(wires=[wires[2], wires[1]])
    return [qml.expval(qml.PauliZ(wires=w)) for w in wires]

wires =np.array(['a', 'c', 'd'], requires_grad=False)
circuit(wires)
```

File: tests/circuit_drawer/test_circuit_drawer.py [hunk 5]
```diff
@@ -603,6 +602,38 @@ def drawn_qubit_circuit_with_probs():
...
+
+@pytest.fixture
+def drawn_qubit_circuit_with_interesting_wires():
+    """The rendered circuit representation of the above qubit circuit."""
+    return (
+        "  0: ──X──╭X──┤ ⟨Y⟩ \n"
+        + " q2: ─────├C──┤     \n"
+        + " -1: ──X──│───┤     \n"
+        + "  b: ──X──╰C──┤     \n"
+    )
```
Comment (josh146) [inline tests/circuit_drawer/test_circuit_drawer.py:634]: 💯

File: tests/circuit_graph/test_circuit_graph.py (hunk 2 above)
Comment (josh146) [reply tests/circuit_graph/test_circuit_graph.py:68]: Does this need to be a fixture? It seems more like test data

File: tests/plugins/test_default_qubit_tf.py [hunk 6]
```diff
@@ -1237,3 +1237,55 @@ def cost(weights):
...
+            qml.CNOT(wires=[wires[2], wires[1]])
+            return qml.probs(wires=wires)
+
+        return circuit
+
+    def make_circuit_expval(self, wires):
+        """Factory for a qnode returning expvals using arbitrary wire labels."""
+        dev = qml.device("default.qubit.tf", wires=wires)
+
+        @qml.qnode(dev)
```
Comment (josh146) [inline tests/plugins/test_default_qubit_tf.py:1263]: interface="tf"?

File: tests/test_qubit_device.py [hunk 7]
```diff
@@ -609,7 +611,7 @@ def apply_over_axes_mock(x, y, p):
 
     marginal_test_data = [
         (np.array([0.1, 0.2, 0.3, 0.4]), np.array([0.4, 0.6]), [1]),
-        (np.array([0.1, 0.2, 0.3, 0.4]), np.array([0.3, 0.7]), [0]),
+        (np.array([0.1, 0.2, 0.3, 0.4]), np.array([0.3, 0.7]), Wires([0])),
```
Comment (josh146) [inline tests/test_qubit_device.py:614]: nice tests!

File: tests/circuit_graph/test_circuit_graph.py (hunk 2 above)
Comment (mariaschuld) [reply tests/circuit_graph/test_circuit_graph.py:68]: You are right!

Comment (mariaschuld) [reply tests/circuit_graph/test_circuit_graph.py:68]: But if it is test data, can I feed it into the fixture ``parametrized_circuit(wires)``? I am currently rather confused about this...

File: tests/beta/test_default_tensor_tf.py [hunk 8]
```diff
@@ -758,3 +758,31 @@ def circuit(x, w=None):
...
+
+
+class TestWiresIntegration:
+    """Test that the device integrates with PennyLane's wire management."""
+
+    def make_circuit_expval(self, wires):
+        """Factory for a qnode returning expvals using arbitrary wire labels."""
+        dev = qml.device("default.tensor.tf", wires=wires)
+
+        @qml.qnode(dev)
```
[the remark this answers is not in the document]
Comment (mariaschuld) [reply tests/beta/test_default_tensor_tf.py:770]: Good point!

[the remark this answers is not in the document]
Comment (mariaschuld) [reply tests/beta/test_default_tensor_tf.py:770]: But the other tests in that file seem to not specify the interface. @co9olguy , do you know why? Can the default.tensor.tf device be used with any interface qnode?

File: tests/beta/test_default_tensor.py (hunk 4 above)
Comment (mariaschuld) [reply tests/beta/test_default_tensor.py:1784]: Seems to work. Full example:

```python
import pennylane as qml
from pennylane import numpy as np

dev = qml.device('default.qubit', wires=['a', 'c', 'd'])

@qml.qnode(dev)
def circuit(wires):
    qml.RX(0.5, wires=wires[0])
    qml.RY(2., wires=wires[1])
    qml.CNOT(wires=[wires[0], wires[2]])
    qml.CNOT(wires=[wires[2], wires[1]])
    return [qml.expval(qml.PauliZ(wires=w)) for w in wires]

wires =np.array(['a', 'c', 'd'], requires_grad=False)
print(circuit(wires))
>>> [ 0.87758256 -0.36520321  0.87758256]
```

Comment (mariaschuld) [reply tests/beta/test_default_tensor.py:1784]: Should we use this in the tests? The reason I didn't is that the device also needs the correct wires...

File: tests/plugins/test_default_qubit_tf.py (hunk 6 above)
Comment (mariaschuld) [reply tests/plugins/test_default_qubit_tf.py:1263]: See above.

File: tests/circuit_graph/test_circuit_graph.py (hunk 2 above)
Comment (josh146) [reply tests/circuit_graph/test_circuit_graph.py:68]: yep, the fixture can use global/module variables in its definition

File: tests/beta/test_default_tensor.py (hunk 4 above)
Comment (josh146) [reply tests/beta/test_default_tensor.py:1784]: Sweet! I don't think it needs to be tested (probably it's already being tested by the QNode integration tests), was just curious

File: tests/beta/test_default_tensor_tf.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (josh146) [reply tests/beta/test_default_tensor_tf.py:770]: @mariaschuld yep it can be used with any interface. If used with autograd or pytorch, it defaults to parameter-shift, if used with tf, it uses backprop (same as `default.qubit.tf`).

File: .github/CHANGELOG.md [hunk 9]
```diff
@@ -2,6 +2,28 @@
 
 <h3>New features since last release</h3>
 
+* The user can now specify custom wire labels, such as `['anc1', 'anc2', 0, 1, 3]`, of type
+  `str` or `Number`. For this, pass a list to the wires argument when creating the device:
+  
+  ```pycon
+  >>> dev = qml.device("default.qubit", wires=['anc1', 'anc2', 0, 1, 3])
```
[the remark this answers is not in the document]
Comment (mariaschuld) [reply .github/CHANGELOG.md:9]: This changelog entry should have been added to the base branch, sorry.

File: tests/beta/test_default_tensor.py [hunk 10]
```diff
@@ -1754,3 +1754,39 @@ def test_hermitian(self, theta, phi, varphi, monkeypatch, rep, tol):
             )
         ) / 16
         assert np.allclose(var, expected, atol=tol, rtol=0)
+
+
+class TestWiresIntegration:
```
Comment (mariaschuld) [inline tests/beta/test_default_tensor.py:1759]: This is a class we could easily reuse for other devices.

File: tests/conftest.py [hunk 11]
```diff
@@ -26,6 +26,7 @@
 TOL = 1e-3
```
Comment (mariaschuld) [inline tests/conftest.py:26]: Must have blacked conftest accidentally?

File: tests/beta/test_default_tensor_tf.py [hunk 12]
```diff
@@ -758,3 +758,39 @@ def circuit(x, w=None):
 
         with pytest.raises(ValueError, match="Device default.tensor.tf only supports diff_method='backprop' when using the tf interface"):
             res = cost_raising_error(params)
+
+
+class TestWiresIntegration:
+    """Test that the device integrates with PennyLane's wire management."""
+
+    def make_circuit_expval(self, wires, representation):
```
Comment (mariaschuld) [inline tests/beta/test_default_tensor_tf.py:766]: Probs is not supported by this device, therefore only testing expval. (I am not sure if both is always necessary, but probs measurements involve a little bit of pipeline that expval does not... )

**Comment by mariaschuld** [comment]
@co9olguy Would you mind having a look? Any other places where we could test wires except from in the devices and circuitgraph/drawer and templates?

File: pennylane/beta/plugins/default_tensor_tf.py [hunk 13]
```diff
@@ -235,14 +235,12 @@ def pre_apply(self):
...
 
         for operation in self.op_queue:
-            # get indices of wires on the device's register
-            registers = self.wire_map(operation.wires)
 
             # Apply each operation, but instead of passing operation.parameters
             # (which contains the evaluated numeric parameter values),
             # pass op_params[operation], which contains numeric values
             # for fixed parameters, and tf.Variable objects for free parameters.
-            super().apply(operation.name, registers, self.op_params[operation])
```
Comment (mariaschuld) [inline pennylane/beta/plugins/default_tensor_tf.py:245]: Another bug uncovered: The apply function actually needs wires and extracts the register...

File: tests/conftest.py (hunk 11 above)
Comment (josh146) [reply tests/conftest.py:26]: Slowly, the entire test folder will be blacked over time 😆

File: tests/beta/test_default_tensor_tf.py (hunk 12 above)
Comment (josh146) [reply tests/beta/test_default_tensor_tf.py:766]: iirc, the only extra pipeline for a device to support `return probs` is for the `Device.probability()` method be defined. This is defined and inherited automatically by devices inheriting `QubitDevice`, but not for `Device`.

Comment (mariaschuld) [reply tests/beta/test_default_tensor_tf.py:766]: Do you think I should remove probability tests altogether and just have a single integration test using expval? In some sense, the rest should be tested in the ``device`` and ``qubit_device`` tests...

**Comment by co9olguy** [comment]
> @co9olguy Would you mind having a look? Any other places where we could test wires except from in the devices and circuitgraph/drawer and templates?

Sorry, didn't find time to get to this yet, it was on my radar for sometime this week though :(
